feat(dashboards) Adapt API to new behavior

markstory opened the pull request:
We're reviving custom dashboards, and need to adapt them to work with discover2 style queries.

* Rename Widget to DashboardWidget as Widget is a common term and using  it here unprefixed will make future us sad.
* Rename WidgetDataSource to DashboardWidgetQuery. There is no  flexibility in widget datasources now. They will all use discover.
* Hide order attributes for widgets and queries. The order in the API  request/response is the implict order. We don't need additional API  parameters.
* Add `interval` to queries so that users can pick bin sizes for their  charts.

There is still some work to be done around validating widgets and their queries but I wanted to do those separately.

I'll do another pull request to remove the old tables once these changes are out. I wanted to avoid potential release boundary problems to be cautious.

These endpoints are unused in production and should be safe to change without consequence.

markstory committed:
feat(dashboards) Adapt API to new behavior

We're reviving custom dashboards, and need to adapt them to work with
discover2 style queries.

* Rename Widget to DashboardWidget as Widget is a common term and using
  it here unprefixed will make future us sad.
* Rename WidgetDataSource to DashboardWidgetQuery. There is no
  flexibility in widget datasources now. They will all use discover.
* Hide order attributes for widgets and queries. The order in the API
  request/response is the implict order. We don't need additional API
  parameters.
* Add `interval` to queries so that users can pick bin sizes for their
  charts.

There is still some work to be done around validating widgets and their
queries but I wanted to do those separately.

These endpoints are unused in production and should be safe to change
without consequence.

markstory commented on src/sentry/migrations/0128_change_dashboards.py:
```diff
@@ -0,0 +1,103 @@
+    operations = [
+        migrations.SeparateDatabaseAndState(
+            database_operations=[
+                migrations.CreateModel(
```
This isn't quite right yet, going to rework this migration.

markstory committed:
Update deletions test.

dashed commented on src/sentry/api/serializers/models/organization_dashboard.py:
```diff
@@ -53,7 +58,9 @@ class DashboardWithWidgetsSerializer(Serializer):
-        widgets = serialize(list(Widget.objects.filter(dashboard_id__in=[i.id for i in item_list])))
+        widgets = serialize(
+            list(DashboardWidget.objects.filter(dashboard_id__in=[i.id for i in item_list]))
+        )
```
It'll be nice to have this pre-sorted by `order` so that we don't have to sort it on the client-side.

markstory replied:
Good idea. I'll also remove the order field. We can use the order of the widgets as the 'order' and not need to maintain additional state on the client.

Zylphrex commented on src/sentry/api/serializers/rest_framework/dashboard_widget.py:
```diff
@@ -0,0 +1,40 @@
+    )
+
+    def validate_displayType(self, display_type):
+        if display_type not in DashboardWidgetDisplayTypes.TYPE_NAMES:
```
Would a choice field be a better choice here?

markstory committed:
Don't use an order parameter

Use the order of the elements in the list instead of an additional order
parameter.

markstory committed:
Use ChoiceField.
